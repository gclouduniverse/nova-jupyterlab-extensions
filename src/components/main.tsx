import {Widget} from '@phosphor/widgets';
import * as React from 'react';

import {GcpService, ProjectState} from '../service/gcp';
import {ReactElementWidget} from '@jupyterlab/apputils';
import {Initializer} from './initializer';
import {INotebookModel} from '@jupyterlab/notebook';

/** Props definition including the GcpService for making backend calls. */
export interface PropsWithGcpService {
  gcpService: GcpService;
}

interface Props extends PropsWithGcpService {
  notebook: INotebookModel;
}

interface State {
  projectState?: ProjectState;
}

export const CSS_BASE = 'jp-MainWidget';

/**
 * Main React component to manage the state of the Scheduler Extension
 */
export class MainWidget extends React.Component<Props, State> {

  constructor(props: Props) {
    super(props);

    this.state = {};
    this._setProjectState = this._setProjectState.bind(this);
  }

  async componentDidMount() {
    this._setProjectState();
  }

  render() {
    const {projectState} = this.state;
    return (<div className={CSS_BASE}>
      <main className={`${CSS_BASE}-main`}>
        {!projectState && <p>Validating project configuration...</p>}
        {projectState && projectState.ready && <p>PLACEHOLDER FOR SCHEDULER</p>}
        {projectState && !projectState.ready
          && <Initializer gcpService={this.props.gcpService}
            projectState={projectState}
            onStateChange={this._setProjectState} />
        }
      </main>
    </div>);
  }

  private async _setProjectState() {
    const projectState = await this.props.gcpService.getProjectState();
    this.setState({projectState});
  }
}

export function mainWidgetFactory(gcpService: GcpService,
  notebook: INotebookModel): Widget {
  const props = {gcpService, notebook};
  return new ReactElementWidget(<MainWidget {...props} />);
}
