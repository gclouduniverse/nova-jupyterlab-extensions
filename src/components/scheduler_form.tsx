import {INotebookModel} from '@jupyterlab/notebook';
import * as React from 'react';
import {withFormik, FormikProps, FormikBag} from 'formik';

import {
  CONTAINER_IMAGES,
  CUSTOM,
  MASTER_TYPES,
  REGIONS,
  SCALE_TIERS,
  SCHEDULE_TYPES,
  SINGLE,
  Option,
} from '../data';
import {RunNotebookRequest} from '../service/gcp';
import {css} from '../styles';
import {CheckboxInput} from './shared/checkbox_input';
import {LearnMoreLink} from './shared/learn_more_link';
import {SchedulerDescription} from './shared/scheduler_description';
import {SelectInput} from './shared/select_input';
import {SubmitButton} from './shared/submit_button';
import {TextInput} from './shared/text_input';
import {PropsWithGcpService, OnDialogClose} from './dialog';

export interface SchedulerFormProps extends PropsWithGcpService {
  notebookName: string;
  notebook: INotebookModel;
  onDialogClose: OnDialogClose;
}

interface SchedulerFormValues {
  jobId: string;
  region: string;
  scaleTier: string;
  masterType?: string;
  imageUri: string;
  scheduleType: string;
  schedule?: string;
  shouldSubmitFiles?: boolean;
}

interface EnhancedSchedulerFormProps extends SchedulerFormProps, FormikProps<SchedulerFormValues> {}

interface SchedulerFormState {
  masterTypesOptions: Option[];
}

const CHECKBOX_LABEL = 'Submit all the files in the with the notebook';
const SCHEDULE_LINK =
  'https://cloud.google.com/scheduler/docs/configuring/cron-job-schedules';
const SCALE_TIER_LINK =
  'https://cloud.google.com/ml-engine/docs/machine-types#scale_tiers';

/** Form Component to submit Scheduled Notebooks */
export class SchedulerForm extends React.Component<EnhancedSchedulerFormProps, SchedulerFormState> {

  constructor(props: EnhancedSchedulerFormProps) {
    super(props);

    this.state = {
      masterTypesOptions: MASTER_TYPES,
    };
  }

  onScaleTierChanged = (e: React.ChangeEvent<any>) => {
    const {handleChange, setFieldValue} = this.props;
    setFieldValue('masterType', e.target.value !== CUSTOM ? '' : MASTER_TYPES[0].value);
    handleChange(e)
  }

  render() {
    const {values, submitForm, handleChange, isSubmitting, status} = this.props;
    return (
      <div>
        <SchedulerDescription />
        <p className={css.bold}>Notebook: {this.props.notebookName}</p>

        <CheckboxInput
          label={CHECKBOX_LABEL}
          name="shouldSubmitFiles"
          onChange={handleChange} />
        <TextInput
          label="Run name"
          name="jobId"
          value={values.jobId}
          onChange={handleChange} />
        <SelectInput
          label="Region"
          name="region"
          value={values.region}
          options={REGIONS}
          onChange={handleChange} />
        <SelectInput
          label="Scale tier"
          name="scaleTier"
          value={values.scaleTier}
          options={SCALE_TIERS}
          onChange={this.onScaleTierChanged} />
        <p className={css.noTopMargin}>
          A scale is a predefined cluster specification.
          <LearnMoreLink href={SCALE_TIER_LINK} />
        </p>
        {values.scaleTier === CUSTOM &&
          <SelectInput
            label="Machine type"
            name="masterType"
            value={values.masterType}
            options={MASTER_TYPES}
            onChange={handleChange} />
        }
        <SelectInput
          label="Container"
          name="imageUri"
          value={values.imageUri}
          options={CONTAINER_IMAGES}
          onChange={handleChange} />
        <SelectInput
          label="Type"
          name="scheduleType"
          value={values.scheduleType}
          options={SCHEDULE_TYPES}
          onChange={handleChange} />
        {values.scheduleType !== SINGLE &&
          <div>
            <TextInput
              label="Frequency"
              name="schedule"
              value={values.schedule}
              onChange={handleChange} />
            <p className={css.noTopMargin}>
              Schedule is specified using unix-cron format. You can define a
              schedule so that your job runs multiple times a day,
              or runs on specific days and months.
              <LearnMoreLink href={SCHEDULE_LINK} />
            </p>
          </div>
        }
        <div className={css.actionBar}>
          <button className={css.button} onClick={this.props.onDialogClose}>
            Cancel
          </button>
          <SubmitButton
            actionPending={isSubmitting}
            onClick={() => {
              submitForm();
            }}
            text='Submit' />
        </div>
        {status && status.msg && <div>{status.msg}</div>}
      </div>
    );
  }
}

export const EnhancedSchedulerForm = withFormik<SchedulerFormProps, SchedulerFormValues>({
  mapPropsToValues: () => (
    {
      jobId: '',
      region: String(REGIONS[1].value),
      scaleTier: String(SCALE_TIERS[0].value),
      imageUri: String(CONTAINER_IMAGES[0].value),
      scheduleType: SINGLE,
      schedule: '',
    }
  ),
  handleSubmit: submit,
})(SchedulerForm);

async function submit(values: SchedulerFormValues, formikBag: FormikBag<SchedulerFormProps, SchedulerFormValues>) {
  const {gcpService, notebook, notebookName} = formikBag.props;
  const {setStatus, setSubmitting} = formikBag;
  const request = await _buildRunNotebookRequest(values, formikBag);

  let status =
    `Uploading ${notebookName} to ${request.inputNotebookGcsPath}`;
  setStatus({msg: status});
  await gcpService.uploadNotebook(notebook.toString(),
    request.inputNotebookGcsPath);

  if (values.scheduleType !== SINGLE && values.schedule) {
    status = 'Submitting Job to Cloud Scheduler';
    setStatus({msg: status});
    // TODO: Obtain Cloud Function URL and Service Account from settings
    const job = await gcpService.scheduleNotebook(request,
      'https://us-central1-prodonjs-kubeflow-dev.cloudfunctions.net/submitScheduledNotebook',
      'prodonjs-kubeflow-dev@appspot.gserviceaccount.com',
      values.schedule
    );
    status = `Successfully created ${job.name}`;
  } else {
    status = 'Submitting Job to AI Platform';
    setStatus({msg: status});
    const job = await gcpService.runNotebook(request);
    status = `Successfully created ${job.jobId}`;
  }
  setStatus({msg: status});
  setSubmitting(false);
}

async function _buildRunNotebookRequest(values: SchedulerFormValues, formikBag: FormikBag<SchedulerFormProps, SchedulerFormValues>): Promise<RunNotebookRequest> {
  const {notebookName} = formikBag.props;
  const {jobId, region, scaleTier, masterType, imageUri} = values;

  // TODO: Obtain bucket from project settings stored during initialization
  const inputNotebookGcsPath = `gs://prodonjs-kubeflow-dev/notebooks/${notebookName}`;
  const outputNotebookGcsPath = inputNotebookGcsPath + '__out.ipynb';

  return {
    jobId,
    imageUri,
    inputNotebookGcsPath,
    masterType: scaleTier !== CUSTOM ? '' : masterType,
    outputNotebookGcsPath,
    scaleTier,
    region,
  };
}
