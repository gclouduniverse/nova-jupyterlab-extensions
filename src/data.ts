/** Helpers for UI selections */
const GCR_PREFIX = 'gcr.io/deeplearning-platform-release';

/** Custom ScaleTier allows selection of AI Platform Machine type */
export const CUSTOM = 'CUSTOM';

/** Indicates a single Notebook run */
export const SINGLE = 'single';
/** Indiactes a recurring scheduled Notebook run */
export const RECURRING = 'recurring';

/** Interface for an <option> inside a <select> */
export interface Option {
  text: string;
  value: string | number;
  disabled?: boolean;
}

/** Returns an option whose value matches the given value. */
export function findOptionByValue<T extends Option>(
  options: T[],
  value: string | number | undefined
): T | undefined {
  if (value === undefined) return undefined;
  return options.find(option => option.value === value);
}

/**
 * Scale tier values for AI Platform Jobs
 * https://cloud.google.com/ml-engine/reference/rest/v1/projects.jobs#scaletier
 */
export const CONTAINER_IMAGES: Option[] = [
  { value: `${GCR_PREFIX}/base-cpu:latest`, text: 'Python' },
  {
    value: `${GCR_PREFIX}/tf-cpu.1-14:latest`,
    text: 'TensorFlow 1.14 (CPU only)',
  },
  { value: `${GCR_PREFIX}/tf-gpu.1-14:latest`, text: 'TensorFlow 1.14 (GPU)' },
  {
    value: `${GCR_PREFIX}/tf2-cpu.2-0:latest`,
    text: 'TensorFlow 2.0 (CPU only)',
  },
  { value: `${GCR_PREFIX}/tf2-gpu.2-0:latest`, text: 'TensorFlow 2.0 (GPU)' },
  {
    value: `${GCR_PREFIX}/pytorch-cpu.1-2:latest`,
    text: 'PyTorch 1.2 (CPU only)',
  },
  { value: `${GCR_PREFIX}/pytorch-gpu.1-2:latest`, text: 'PyTorch 1.2 (GPU)' },
  {
    value: `${GCR_PREFIX}/r-cpu.3-6:latest`,
    text: 'R 3.6 (with r-essentials)',
  },
];

/**
 * Scale tier values for AI Platform Jobs
 * https://cloud.google.com/ml-engine/reference/rest/v1/projects.jobs#scaletier
 */
export const SCALE_TIERS: Option[] = [
  { value: 'BASIC', text: 'Single worker instance' },
  {
    value: 'STANDARD_1',
    text: 'Many worker instances and a few parameter servers',
  },
  {
    value: 'PREMIUM_1',
    text: 'A large number of workers with many parameter servers',
  },
  { value: 'BASIC_GPU', text: 'A single worker instance with a GPU' },
  { value: 'BASIC_TPU', text: 'A single worker instance with a Cloud TPU' },
  { value: CUSTOM, text: 'Custom machine type configuration' },
];

/**
 * AI Platform Machine types.
 * https://cloud.google.com/ml-engine/docs/machine-types#compare-machine-types
 */
export const MASTER_TYPES: Option[] = [
  { value: 'standard', text: '4 CPUs, 15 GB RAM, No Accelerators' },
  { value: 'large_model', text: '8 CPUs, 52 GB RAM, No Accelerators' },
  { value: 'complex_model_s', text: '8 CPUs, 7.2 GB RAM, No Accelerators' },
  { value: 'complex_model_m', text: '16 CPUs, 14.4 GB RAM, No Accelerators' },
  { value: 'complex_model_l', text: '32 CPUs, 28.8 GB RAM, No Accelerators' },
  { value: 'standard_gpu', text: '8 CPUs, 30 GB RAM, 1 NVIDIA Tesla K80 GPU' },
  {
    value: 'complex_model_m_gpu',
    text: '4 CPUs, 15 GB RAM, 4 NVIDIA Tesla K80 GPUs',
  },
  {
    value: 'complex_model_l_gpu',
    text: '4 CPUs, 15 GB RAM, 8 NVIDIA Tesla K80 GPUs',
  },
  {
    value: 'standard_p100',
    text: '8 CPUs, 30 GB RAM, 1 NVIDIA Tesla P100 GPU',
  },
  {
    value: 'complex_model_m_p100',
    text: '4 CPUs, 15 GB RAM, 4 NVIDIA Tesla K80 GPUs',
  },
  {
    value: 'complex_model_l_gpu',
    text: '4 CPUs, 15 GB RAM, 8 NVIDIA Tesla K80 GPUs',
  },
  { value: 'cloud_tpu', text: '8 TPU cores' },
];

/**
 * Supported AI Platform regions.
 * https://cloud.google.com/ml-engine/docs/regions
 * TODO: It may be more sensible to invoke the projects.locations.list API
 * and filter for locations with TRAINING capability
 * https://cloud.google.com/ml-engine/reference/rest/v1/projects.locations/list
 */
export const REGIONS: Option[] = [
  {
    value: 'us-central1',
    text: 'us-central1 (Iowa)',
  },
  {
    value: 'us-east1',
    text: 'us-east1 (South Carolina)',
  },
  {
    value: 'us-west1',
    text: 'us-west1 (Oregon)',
  },
  {
    value: 'asia-east1',
    text: 'asia-east1 (Taiwan)',
  },
  {
    value: 'europe-west1',
    text: 'europe-west1 (Belgium)',
  },
  {
    value: 'europe-west4',
    text: 'europe-west4 (Netherlands)',
  },
  {
    value: 'asia-east1',
    text: 'asia-east1 (Taiwan)',
  },
  {
    value: 'asia-southeast1',
    text: 'asia-southeast1 (Singapore)',
  },
];

/**
 * Supported AppEngine locations, which dictate the endpoint for Cloud Scheduler
 * https://cloud.google.com/appengine/docs/locations
 * Because we deploy the Cloud Function to the same region, this is actually
 * an intersection of the above list and the one at
 * https://cloud.google.com/functions/docs/locations
 */
export const APPENGINE_REGIONS: Option[] = [
  {
    value: 'us-central1',
    text: 'us-central1 (Iowa)',
  },
  {
    value: 'us-east1',
    text: 'us-east1 (South Carolina)',
  },
  {
    value: 'us-east4',
    text: 'us-east4 (Northern Virginia)',
  },
  {
    value: 'asia-east2',
    text: 'asia-east2 (Hong Kong)',
  },
  {
    value: 'asia-northeast1',
    text: 'asia-northeast1 (Tokyo)',
  },
  {
    value: 'europe-west1',
    text: 'europe-west1 (Belgium)',
  },
  {
    value: 'europe-west2',
    text: 'europe-west2 (London)',
  },
];

/** Single execution or recurring schedule */
export const SCHEDULE_TYPES: Option[] = [
  { value: SINGLE, text: 'Single run' },
  { value: RECURRING, text: 'Recurring run' },
];
