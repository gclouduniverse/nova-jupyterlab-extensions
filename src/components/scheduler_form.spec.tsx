import * as React from 'react';
import {shallow} from 'enzyme';

import {GcpService, RunNotebookRequest} from '../service/gcp';
import {MainProps} from './main';
import {SchedulerForm} from './scheduler_form';
import {CUSTOM, CONTAINER_IMAGES} from '../data';
import {INotebookModel} from '@jupyterlab/notebook';

const CHANGE = 'change';

describe('SchedulerForm', () => {
  const notebookName = 'Test Notebook.ipynb';
  const notebookContents = '{"notebook": "test"}';
  const gcsPath = `gs://prodonjs-kubeflow-dev/notebooks/${notebookName}`;

  const mockUploadNotebook = jest.fn();
  const mockRunNotebook = jest.fn();
  const mockScheduleNotebook = jest.fn();
  const mockGcpService = {
    uploadNotebook: mockUploadNotebook,
    runNotebook: mockRunNotebook,
    scheduleNotebook: mockScheduleNotebook
  } as undefined as GcpService;
  const mockNotebookContents = jest.fn();
  const mockProps: MainProps = {
    gcpService: mockGcpService,
    notebook: {
      toString: mockNotebookContents
    } as unknown as INotebookModel,
    notebookName,
  };

  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('Toggles Machine type visibility based on Scale tier', async () => {
    const schedulerForm = shallow(<SchedulerForm {...mockProps} />);
    expect(schedulerForm).toMatchSnapshot('Without MachineTypes');
    schedulerForm.find('SelectInput[label="Scale tier"]')
      .simulate(CHANGE, CUSTOM);

    expect(schedulerForm).toMatchSnapshot('With MachineTypes');
  });

  it('Toggles Schedule visibility based on Frequency', async () => {
    const schedulerForm = shallow(<SchedulerForm {...mockProps} />);
    expect(schedulerForm).toMatchSnapshot('Without Schedule');

    schedulerForm.find('SelectInput[label="Type"]')
      .simulate(CHANGE, 'recurring');
    expect(schedulerForm).toMatchSnapshot('With Schedule');
  });

  it('Submits an immediate job to AI Platform', async () => {
    const uploadNotebookPromise = Promise.resolve();
    const runNotebookPromise = Promise.resolve({jobId: 'aiplatform_job_1'});

    mockNotebookContents.mockReturnValue(notebookContents);
    mockUploadNotebook.mockReturnValue(uploadNotebookPromise);
    mockRunNotebook.mockReturnValue(runNotebookPromise);

    const schedulerForm = shallow(<SchedulerForm {...mockProps} />);
    schedulerForm.find('TextInput[label="Run name"]')
      .simulate(CHANGE, 'test_immediate_job');
    schedulerForm.find('SubmitButton').simulate('click');
    expect(schedulerForm
      .contains(<p>Uploading {notebookName} to {gcsPath}</p>)).toBe(true);

    await uploadNotebookPromise;

    expect(schedulerForm
      .contains(<p>Submitting Job to AI Platform</p>)).toBe(true);
    await runNotebookPromise;

    expect(schedulerForm
      .contains(<p>Successfully created aiplatform_job_1</p>)).toBe(true);
    expect(mockGcpService.uploadNotebook)
      .toHaveBeenCalledWith(notebookContents, gcsPath);
    const aiPlatformRequest: RunNotebookRequest = {
      jobId: 'test_immediate_job',
      imageUri: 'gcr.io/deeplearning-platform-release/base-cpu:latest',
      inputNotebookGcsPath: gcsPath,
      masterType: undefined,
      outputNotebookGcsPath: `${gcsPath}__out.ipynb`,
      scaleTier: 'BASIC',
      region: 'us-central1',
    };
    expect(mockGcpService.runNotebook).toHaveBeenCalledWith(aiPlatformRequest);
  });

  it('Submits a scheduled job to Cloud Scheduler', async () => {
    const uploadNotebookPromise = Promise.resolve();
    const scheduleNotebookPromise = Promise.resolve(
      {name: 'cloudscheduler_job_1'});

    mockNotebookContents.mockReturnValue(notebookContents);
    mockUploadNotebook.mockReturnValue(uploadNotebookPromise);
    mockScheduleNotebook.mockReturnValue(scheduleNotebookPromise);

    const schedulerForm = shallow(<SchedulerForm {...mockProps} />);
    schedulerForm.find('TextInput[label="Run name"]')
      .simulate(CHANGE, 'test_scheduled_job');
    schedulerForm.find('SelectInput[label="Region"]')
      .simulate(CHANGE, 'us-east1');
    schedulerForm.find('SelectInput[label="Scale tier"]')
      .simulate(CHANGE, CUSTOM);
    schedulerForm.find('SelectInput[label="Machine type"]')
      .simulate(CHANGE, 'complex_model_m_gpu');
    schedulerForm.find('SelectInput[label="Container"]')
      .simulate(CHANGE, CONTAINER_IMAGES[2].value);
    schedulerForm.find('SelectInput[label="Type"]')
      .simulate(CHANGE, 'recurring');
    schedulerForm.find('TextInput[label="Frequency"]')
      .simulate(CHANGE, '0 0 * * *');
    schedulerForm.find('SubmitButton').simulate('click');

    await uploadNotebookPromise;
    await scheduleNotebookPromise;

    expect(mockGcpService.uploadNotebook)
      .toHaveBeenCalledWith(notebookContents, gcsPath);
    const aiPlatformRequest: RunNotebookRequest = {
      jobId: 'test_scheduled_job',
      imageUri: 'gcr.io/deeplearning-platform-release/tf-gpu.1-14:latest',
      inputNotebookGcsPath: gcsPath,
      masterType: 'complex_model_m_gpu',
      outputNotebookGcsPath: `${gcsPath}__out.ipynb`,
      scaleTier: 'CUSTOM',
      region: 'us-east1',
    };
    expect(mockGcpService.scheduleNotebook).toHaveBeenCalledWith(
      aiPlatformRequest,
      'https://us-central1-prodonjs-kubeflow-dev.cloudfunctions.net/submitScheduledNotebook',
      'prodonjs-kubeflow-dev@appspot.gserviceaccount.com', '0 0 * * *');
  });
});
