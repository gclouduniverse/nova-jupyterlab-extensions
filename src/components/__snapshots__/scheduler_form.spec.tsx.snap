// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchedulerForm Toggles Machine type visibility based on Scale tier: With MachineTypes 1`] = `
<div>
  <SchedulerDescription />
  <p
    className="bold_f1wf24vo"
  >
    Notebook: 
    Test Notebook.ipynb
  </p>
  <CheckboxInput
    label="Submit all the files in the with the notebook"
    onChange={[Function]}
  />
  <TextInput
    label="Run name"
    onChange={[Function]}
  />
  <SelectInput
    label="Region"
    onChange={[Function]}
    options={
      Array [
        Object {
          "text": "us-central1 (Iowa)",
          "value": "us-central1",
        },
        Object {
          "text": "us-east1 (South Carolina)",
          "value": "us-east1",
        },
        Object {
          "text": "us-west1 (Oregon)",
          "value": "us-west1",
        },
        Object {
          "text": "asia-east1 (Taiwan)",
          "value": "asia-east1",
        },
        Object {
          "text": "europe-west1 (Belgium)",
          "value": "europe-west1",
        },
        Object {
          "text": "europe-west4 (Netherlands)",
          "value": "europe-west4",
        },
        Object {
          "text": "asia-east1 (Taiwan)",
          "value": "asia-east1",
        },
        Object {
          "text": "asia-southeast1 (Singapore)",
          "value": "asia-southeast1",
        },
      ]
    }
  />
  <SelectInput
    label="Scale tier"
    onChange={[Function]}
    options={
      Array [
        Object {
          "text": "Single worker instance",
          "value": "BASIC",
        },
        Object {
          "text": "Many worker instances and a few parameter servers",
          "value": "STANDARD_1",
        },
        Object {
          "text": "A large number of workers with many parameter servers",
          "value": "PREMIUM_1",
        },
        Object {
          "text": "A single worker instance with a GPU",
          "value": "BASIC_GPU",
        },
        Object {
          "text": "A single worker instance with a Cloud TPU",
          "value": "BASIC_TPU",
        },
        Object {
          "text": "Custom machine type configuration",
          "value": "CUSTOM",
        },
      ]
    }
  />
  <p
    className="noTopMargin_felu21j"
  >
    A scale is a predefined cluster specification.
    <LearnMoreLink
      href="https://cloud.google.com/ml-engine/docs/machine-types#scale_tiers"
    />
  </p>
  <SelectInput
    label="Machine type"
    onChange={[Function]}
    options={
      Array [
        Object {
          "text": "4 CPUs, 15 GB RAM, No Accelerators",
          "value": "standard",
        },
        Object {
          "text": "8 CPUs, 52 GB RAM, No Accelerators",
          "value": "large_model",
        },
        Object {
          "text": "8 CPUs, 7.2 GB RAM, No Accelerators",
          "value": "complex_model_s",
        },
        Object {
          "text": "16 CPUs, 14.4 GB RAM, No Accelerators",
          "value": "complex_model_m",
        },
        Object {
          "text": "32 CPUs, 28.8 GB RAM, No Accelerators",
          "value": "complex_model_l",
        },
        Object {
          "text": "8 CPUs, 30 GB RAM, 1 NVIDIA Tesla K80 GPU",
          "value": "standard_gpu",
        },
        Object {
          "text": "4 CPUs, 15 GB RAM, 4 NVIDIA Tesla K80 GPUs",
          "value": "complex_model_m_gpu",
        },
        Object {
          "text": "4 CPUs, 15 GB RAM, 8 NVIDIA Tesla K80 GPUs",
          "value": "complex_model_l_gpu",
        },
        Object {
          "text": "8 CPUs, 30 GB RAM, 1 NVIDIA Tesla P100 GPU",
          "value": "standard_p100",
        },
        Object {
          "text": "4 CPUs, 15 GB RAM, 4 NVIDIA Tesla K80 GPUs",
          "value": "complex_model_m_p100",
        },
        Object {
          "text": "4 CPUs, 15 GB RAM, 8 NVIDIA Tesla K80 GPUs",
          "value": "complex_model_l_gpu",
        },
        Object {
          "text": "8 TPU cores",
          "value": "cloud_tpu",
        },
      ]
    }
  />
  <SelectInput
    label="Type"
    onChange={[Function]}
    options={
      Array [
        Object {
          "text": "Single run",
          "value": "single",
        },
        Object {
          "text": "Recurring run",
          "value": "recurring",
        },
      ]
    }
  />
  <div
    className="actionBar_fe1jzmp"
  >
    <button
      className="button_f1o5na2b"
      onClick={[MockFunction]}
    >
      Cancel
    </button>
    <SubmitButton
      actionPending={false}
      onClick={[Function]}
      text="Submit"
    />
  </div>
</div>
`;

exports[`SchedulerForm Toggles Machine type visibility based on Scale tier: Without MachineTypes 1`] = `
<div>
  <SchedulerDescription />
  <p
    className="bold_f1wf24vo"
  >
    Notebook: 
    Test Notebook.ipynb
  </p>
  <CheckboxInput
    label="Submit all the files in the with the notebook"
    onChange={[Function]}
  />
  <TextInput
    label="Run name"
    onChange={[Function]}
  />
  <SelectInput
    label="Region"
    onChange={[Function]}
    options={
      Array [
        Object {
          "text": "us-central1 (Iowa)",
          "value": "us-central1",
        },
        Object {
          "text": "us-east1 (South Carolina)",
          "value": "us-east1",
        },
        Object {
          "text": "us-west1 (Oregon)",
          "value": "us-west1",
        },
        Object {
          "text": "asia-east1 (Taiwan)",
          "value": "asia-east1",
        },
        Object {
          "text": "europe-west1 (Belgium)",
          "value": "europe-west1",
        },
        Object {
          "text": "europe-west4 (Netherlands)",
          "value": "europe-west4",
        },
        Object {
          "text": "asia-east1 (Taiwan)",
          "value": "asia-east1",
        },
        Object {
          "text": "asia-southeast1 (Singapore)",
          "value": "asia-southeast1",
        },
      ]
    }
  />
  <SelectInput
    label="Scale tier"
    onChange={[Function]}
    options={
      Array [
        Object {
          "text": "Single worker instance",
          "value": "BASIC",
        },
        Object {
          "text": "Many worker instances and a few parameter servers",
          "value": "STANDARD_1",
        },
        Object {
          "text": "A large number of workers with many parameter servers",
          "value": "PREMIUM_1",
        },
        Object {
          "text": "A single worker instance with a GPU",
          "value": "BASIC_GPU",
        },
        Object {
          "text": "A single worker instance with a Cloud TPU",
          "value": "BASIC_TPU",
        },
        Object {
          "text": "Custom machine type configuration",
          "value": "CUSTOM",
        },
      ]
    }
  />
  <p
    className="noTopMargin_felu21j"
  >
    A scale is a predefined cluster specification.
    <LearnMoreLink
      href="https://cloud.google.com/ml-engine/docs/machine-types#scale_tiers"
    />
  </p>
  <SelectInput
    label="Type"
    onChange={[Function]}
    options={
      Array [
        Object {
          "text": "Single run",
          "value": "single",
        },
        Object {
          "text": "Recurring run",
          "value": "recurring",
        },
      ]
    }
  />
  <div
    className="actionBar_fe1jzmp"
  >
    <button
      className="button_f1o5na2b"
      onClick={[MockFunction]}
    >
      Cancel
    </button>
    <SubmitButton
      actionPending={false}
      onClick={[Function]}
      text="Submit"
    />
  </div>
</div>
`;

exports[`SchedulerForm Toggles Schedule visibility based on Frequency: With Schedule 1`] = `
<div>
  <SchedulerDescription />
  <p
    className="bold_f1wf24vo"
  >
    Notebook: 
    Test Notebook.ipynb
  </p>
  <CheckboxInput
    label="Submit all the files in the with the notebook"
    onChange={[Function]}
  />
  <TextInput
    label="Run name"
    onChange={[Function]}
  />
  <SelectInput
    label="Region"
    onChange={[Function]}
    options={
      Array [
        Object {
          "text": "us-central1 (Iowa)",
          "value": "us-central1",
        },
        Object {
          "text": "us-east1 (South Carolina)",
          "value": "us-east1",
        },
        Object {
          "text": "us-west1 (Oregon)",
          "value": "us-west1",
        },
        Object {
          "text": "asia-east1 (Taiwan)",
          "value": "asia-east1",
        },
        Object {
          "text": "europe-west1 (Belgium)",
          "value": "europe-west1",
        },
        Object {
          "text": "europe-west4 (Netherlands)",
          "value": "europe-west4",
        },
        Object {
          "text": "asia-east1 (Taiwan)",
          "value": "asia-east1",
        },
        Object {
          "text": "asia-southeast1 (Singapore)",
          "value": "asia-southeast1",
        },
      ]
    }
  />
  <SelectInput
    label="Scale tier"
    onChange={[Function]}
    options={
      Array [
        Object {
          "text": "Single worker instance",
          "value": "BASIC",
        },
        Object {
          "text": "Many worker instances and a few parameter servers",
          "value": "STANDARD_1",
        },
        Object {
          "text": "A large number of workers with many parameter servers",
          "value": "PREMIUM_1",
        },
        Object {
          "text": "A single worker instance with a GPU",
          "value": "BASIC_GPU",
        },
        Object {
          "text": "A single worker instance with a Cloud TPU",
          "value": "BASIC_TPU",
        },
        Object {
          "text": "Custom machine type configuration",
          "value": "CUSTOM",
        },
      ]
    }
  />
  <p
    className="noTopMargin_felu21j"
  >
    A scale is a predefined cluster specification.
    <LearnMoreLink
      href="https://cloud.google.com/ml-engine/docs/machine-types#scale_tiers"
    />
  </p>
  <SelectInput
    label="Type"
    onChange={[Function]}
    options={
      Array [
        Object {
          "text": "Single run",
          "value": "single",
        },
        Object {
          "text": "Recurring run",
          "value": "recurring",
        },
      ]
    }
  />
  <div>
    <TextInput
      label="Frequency"
      onChange={[Function]}
    />
    <p
      className="noTopMargin_felu21j"
    >
      Schedule is specified using unix-cron format. You can define a schedule so that your job runs multiple times a day, or runs on specific days and months.
      <LearnMoreLink
        href="https://cloud.google.com/scheduler/docs/configuring/cron-job-schedules"
      />
    </p>
  </div>
  <div
    className="actionBar_fe1jzmp"
  >
    <button
      className="button_f1o5na2b"
      onClick={[MockFunction]}
    >
      Cancel
    </button>
    <SubmitButton
      actionPending={false}
      onClick={[Function]}
      text="Submit"
    />
  </div>
</div>
`;

exports[`SchedulerForm Toggles Schedule visibility based on Frequency: Without Schedule 1`] = `
<div>
  <SchedulerDescription />
  <p
    className="bold_f1wf24vo"
  >
    Notebook: 
    Test Notebook.ipynb
  </p>
  <CheckboxInput
    label="Submit all the files in the with the notebook"
    onChange={[Function]}
  />
  <TextInput
    label="Run name"
    onChange={[Function]}
  />
  <SelectInput
    label="Region"
    onChange={[Function]}
    options={
      Array [
        Object {
          "text": "us-central1 (Iowa)",
          "value": "us-central1",
        },
        Object {
          "text": "us-east1 (South Carolina)",
          "value": "us-east1",
        },
        Object {
          "text": "us-west1 (Oregon)",
          "value": "us-west1",
        },
        Object {
          "text": "asia-east1 (Taiwan)",
          "value": "asia-east1",
        },
        Object {
          "text": "europe-west1 (Belgium)",
          "value": "europe-west1",
        },
        Object {
          "text": "europe-west4 (Netherlands)",
          "value": "europe-west4",
        },
        Object {
          "text": "asia-east1 (Taiwan)",
          "value": "asia-east1",
        },
        Object {
          "text": "asia-southeast1 (Singapore)",
          "value": "asia-southeast1",
        },
      ]
    }
  />
  <SelectInput
    label="Scale tier"
    onChange={[Function]}
    options={
      Array [
        Object {
          "text": "Single worker instance",
          "value": "BASIC",
        },
        Object {
          "text": "Many worker instances and a few parameter servers",
          "value": "STANDARD_1",
        },
        Object {
          "text": "A large number of workers with many parameter servers",
          "value": "PREMIUM_1",
        },
        Object {
          "text": "A single worker instance with a GPU",
          "value": "BASIC_GPU",
        },
        Object {
          "text": "A single worker instance with a Cloud TPU",
          "value": "BASIC_TPU",
        },
        Object {
          "text": "Custom machine type configuration",
          "value": "CUSTOM",
        },
      ]
    }
  />
  <p
    className="noTopMargin_felu21j"
  >
    A scale is a predefined cluster specification.
    <LearnMoreLink
      href="https://cloud.google.com/ml-engine/docs/machine-types#scale_tiers"
    />
  </p>
  <SelectInput
    label="Type"
    onChange={[Function]}
    options={
      Array [
        Object {
          "text": "Single run",
          "value": "single",
        },
        Object {
          "text": "Recurring run",
          "value": "recurring",
        },
      ]
    }
  />
  <div
    className="actionBar_fe1jzmp"
  >
    <button
      className="button_f1o5na2b"
      onClick={[MockFunction]}
    >
      Cancel
    </button>
    <SubmitButton
      actionPending={false}
      onClick={[Function]}
      text="Submit"
    />
  </div>
</div>
`;
